Mathlib/FieldTheory/Adjoin.lean

```diff
@@ -1237,45 +1043,139 @@ instance finiteDimensional_iSup_of_finite {ι : Type*} {t : ι → IntermediateF
+  le_trans L₁ L₂ L₃ := by
+    rintro ⟨h₁₂, h₁₂'⟩ ⟨h₂₃, h₂₃'⟩
+    refine ⟨h₁₂.trans h₂₃, fun _ ↦ ?_⟩
+    rw [← inclusion_inclusion h₁₂ h₂₃, h₂₃', h₁₂']
+  le_antisymm := by
+    rintro ⟨L₁, e₁⟩ ⟨L₂, e₂⟩ ⟨h₁₂, h₁₂'⟩ ⟨h₂₁, h₂₁'⟩
+    obtain rfl : L₁ = L₂ := h₁₂.antisymm h₂₁
+    congr
+    exact AlgHom.ext h₂₁'
+
+noncomputable instance : OrderBot (Lifts F E K) where
+  bot := ⟨⊥, (Algebra.ofId F K).comp (botEquiv F E)⟩
+  bot_le L := ⟨bot_le, fun x ↦ by
+    obtain ⟨x, rfl⟩ := (botEquiv F E).symm.surjective x
+    simp_rw [AlgHom.comp_apply, AlgHom.coe_coe, AlgEquiv.apply_symm_apply]
+    exact L.emb.commutes x⟩
+
+noncomputable instance : Inhabited (Lifts F E K) :=
+  ⟨⊥⟩
+
+/-- A chain of lifts has an upper bound. -/
+theorem Lifts.exists_upper_bound (c : Set (Lifts F E K)) (hc : IsChain (· ≤ ·) c) :
+    ∃ ub, ∀ a ∈ c, a ≤ ub := by
+  let t (i : ↑(insert ⊥ c)) := i.val.carrier
+  let t' (i) := (t i).toSubalgebra
+  have hc := hc.insert fun _ _ _ ↦ .inl bot_le
+  have dir : Directed (· ≤ ·) t := hc.directedOn.directed_val.mono_comp _ fun _ _ h ↦ h.1
+  refine ⟨⟨iSup t, (Subalgebra.iSupLift t' dir (fun i ↦ i.val.emb) (fun i j h ↦ ?_) _ rfl).comp
+      (Subalgebra.equivOfEq _ _ <| toSubalgebra_iSup_of_directed dir)⟩,
+    fun L hL ↦ have hL := Set.mem_insert_of_mem ⊥ hL; ⟨le_iSup t ⟨L, hL⟩, fun x ↦ ?_⟩⟩
+  · refine AlgHom.ext fun x ↦ (hc.total i.2 j.2).elim (fun hij ↦ (hij.snd x).symm) fun hji ↦ ?_
+    erw [AlgHom.comp_apply, ← hji.snd (Subalgebra.inclusion h x),
+         inclusion_inclusion, inclusion_self, AlgHom.id_apply x]
+  · dsimp only [AlgHom.comp_apply]
+    exact Subalgebra.iSupLift_inclusion (K := t') (i := ⟨L, hL⟩) x (le_iSup t' ⟨L, hL⟩)
+#align intermediate_field.lifts.exists_upper_bound IntermediateField.Lifts.exists_upper_bound
+
+/-- Given an element `s : E` whose conjugates are all in `K`, any lift can be extended to one
+  whose carrier contains `s`. -/
+theorem Lifts.exists_lift_of_splits (x : Lifts F E K) {s : E} (h1 : IsIntegral F s)
+    (h2 : (minpoly F s).Splits (algebraMap F K)) : ∃ y, x ≤ y ∧ s ∈ y.carrier :=
+  have I1 : IsIntegral x.carrier s := isIntegral_of_isScalarTower h1
+  have I2 := (minpoly.degree_pos I1).ne'
+  have key : (minpoly x.carrier s).Splits x.emb.toRingHom :=
+    splits_of_splits_of_dvd _ (map_ne_zero (minpoly.ne_zero h1))
+      ((splits_map_iff _ _).2 (x.emb.comp_algebraMap ▸ h2)) (minpoly.dvd_map_of_isScalarTower _ _ _)
+  letI : Algebra x.carrier K := x.emb.toRingHom.toAlgebra
+  let carrier := x.carrier⟮s⟯.restrictScalars F
+  letI : Algebra x.carrier carrier := x.carrier⟮s⟯.toSubalgebra.algebra
+  let φ : carrier →ₐ[x.carrier] K := ((algHomAdjoinIntegralEquiv x.carrier I1).symm
+    ⟨rootOfSplits x.emb.toRingHom key I2, by
+      rw [mem_aroots, and_iff_right (minpoly.ne_zero I1)]
+      exact map_rootOfSplits x.emb.toRingHom key I2⟩)
+  ⟨⟨carrier, (@algHomEquivSigma F x.carrier carrier K _ _ _ _ _ _ _ _
+      (IsScalarTower.of_algebraMap_eq fun _ ↦ rfl)).symm ⟨x.emb, φ⟩⟩,
```

Comment:
I'm unable to pull out the `IsScalarTower` instance because Lean can't find the `SMul x.carrier carrier` instance even though the `Algebra` instance is there. I think I encountered this `Algebra`->`SMul` failure elsewhere as well.